Set initial timestamp to the time the image was created

mac-chaffee opened the pull request:
Previously, when docuum first starts, it would set the timestamp of every image to the time docuum started, meaning old images would be removed at random. Now, we use the time the image was created as the initial timestamp (which is of course updated when that image is used like normal).

Example running locally (with comments added to show the original CreatedAt dates):
```
$ cargo run -- --threshold '50 GB'
   Compiling docuum v0.9.5 (/Users/machaffe/personal/docuum)
    Finished dev [unoptimized + debuginfo] target(s) in 3.81s
     Running `target/debug/docuum --threshold '50 GB'`
[2020-07-17 18:20:11 -04:00 INFO] Waking up…
[2020-07-17 18:20:11 -04:00 INFO] Docker images are currently using 75.27 GB but the limit is 50.00 GB. Some images will be deleted.
[2020-07-17 18:20:12 -04:00 INFO] Deleting image sha256:3129a2ca29d75226dc5657a4629cdd5f38accda7f5b75bc5d5a76f5b4e0e5870…
Untagged: quay.io/coreos/configmap-reload:v0.0.1            <--- CreatedAt Feb 2017
Untagged: quay.io/coreos/configmap-reload@sha256:e2fd60ff0ae4500a75b80ebaa30e0e7deba9ad107833e8ca53f0047c42c5a057
Deleted: sha256:3129a2ca29d75226dc5657a4629cdd5f38accda7f5b75bc5d5a76f5b4e0e5870
Deleted: sha256:0065890d4ef6d9e21ba99dea6a990abe8b53db0b82562fe68dd3f37ef529b156
Deleted: sha256:91bd48b9e0b007ce035cccf10d1dc7e482695cf492c94808875fa84e71a85482
[2020-07-17 18:20:12 -04:00 INFO] Deleting image sha256:2faf6f7a322f5f8d565e05675d5ab6f4d1b4a4f65f88cb14df15e76b902b0536…
Untagged: quay.io/coreos/hyperkube:v1.7.6_coreos.0           <--- CreatedAt Sep 2017
Untagged: quay.io/coreos/hyperkube@sha256:699fc03fccb1c4662fee9d996cf30106aea55a6f594d16c1504cc7334dadcee4
Deleted: sha256:2faf6f7a322f5f8d565e05675d5ab6f4d1b4a4f65f88cb14df15e76b902b0536
Deleted: sha256:e5f68041508088cc8a24342143cb911297b424db7cfd8c2f4362124382a9eaf3
Deleted: sha256:3b91c2286ccf87e35ab72d4fbcc66f7b0d6834408c69726d250d778759c75359
Deleted: sha256:713bebfb34f9e43249fd9c44b3c935a0fcdc8575a02822f4fbfc65f6fd6a89d4
Deleted: sha256:c18c4811ea9f56e6561adc47460f2a42937e33b3e6782911ea4a3f36f01438c1
Deleted: sha256:fed386583ce4ea496bad7914fc503d437f8ba8d3590fb7c69ed8c526f879d6c0
Deleted: sha256:18f9b4e2e1bcd5abe381a557c44cba379884c88f6049564f58fd8c10ab5733df
[2020-07-17 18:20:14 -04:00 INFO] Deleting image sha256:da86e6ba6ca197bf6bc5e9d900febd906b133eaa4750e6bed647b0fbe50ed43e…
Error response from daemon: conflict: unable to delete da86e6ba6ca1 (cannot be forced) - image is being used by running container ac477f13429f
[2020-07-17 18:20:14 -04:00 ERROR] Unable to delete image sha256:da86e6ba6ca197bf6bc5e9d900febd906b133eaa4750e6bed647b0fbe50ed43e.
[2020-07-17 18:20:14 -04:00 INFO] Deleting image sha256:c81481184b1b001f225769d1482b360b4be18fe7dd53fa93cdbd45d568e905a7…
Untagged: justincormack/nsenter1:latest                      <--- CreatedAt Jan 2018
Untagged: justincormack/nsenter1@sha256:5af0be5e42ebd55eea2c593e4622f810065c3f45bb805eaacf43f08f3d06ffd8
Deleted: sha256:c81481184b1b001f225769d1482b360b4be18fe7dd53fa93cdbd45d568e905a7
Deleted: sha256:c43e635df3c82c2f4ce61495e8440e23ed6001c7c26c8353e6b743459a97d60e
<truncated>
```

Also this is my first foray into Rust, so please let me know if there are any best-practices I should be following, particularly around error handling 👍 

**Status:** Ready

**Fixes:** #71

mac-chaffee committed:
Set initial timestamp to the time the image was created

Previously, when docuum first starts, it would set the timestamp of every
image to the time docuum started, meaning old images would be removed at
random. Now, we use the time the image was created as the initial timestamp
(which is of course updated when that image is used like normal).

stepchowfun commented:
Thanks so much for contributing! I'll take a look soon.

A small tip: I see you have a small lint error. If you have [Toast](https://github.com/stepchowfun/toast) installed, you can run `toast lint` to reproduce it locally. If you do not want to use Toast, you can run the various linters (e.g., Clippy, ShellCheck) manually or just rely on the CI system we have here.

mac-chaffee committed:
Display error message when cargo fmt step fails

stepchowfun commented on src/run.rs:
```diff
@@ -4,13 +4,22 @@ use crate::{
+    DateTime,
+    Local
+};
+
```
For consistency, ca we remove this empty line and the one two lines down? It seems the convention here is to not have empty lines between the `use` directives.

stepchowfun commented on src/run.rs:
```diff
@@ -95,12 +111,40 @@ pub fn image_ids() -> io::Result<HashSet<String>> {
                 .lines()
                 .map(|line| line.trim().to_owned())
-                .filter(|line| !line.is_empty())
+                .filter_map(|line| {
```
I'm thinking: maybe we can combine this `filter_map` with the `map` directly above it into a single `filter_map`. It would first trim the line, store the result in a new variable (like `trimmed_line` or something like that), and use that variable for the rest of the logic.

stepchowfun commented on src/run.rs:
```diff
@@ -95,12 +111,40 @@ pub fn image_ids() -> io::Result<HashSet<String>> {
-                .filter(|line| !line.is_empty())
+                .filter_map(|line| {
+                    if line.is_empty() {
+                        let tab_index = line.find('\t').unwrap();
```
If for some reason Docker outputs some garbage, it would be strongly preferred to return an `io::Error` over `panic`-ing (which is what `unwrap` would do). Fortunately, the return type already supports this with `io::Result`.

Since Docuum is meant to be a long running process that essentially never crashes (except under extreme circumstances like the system running out of memory), we want to be extremely judicious with any code that might panic. In particular, we generally only `unwrap` in situations where the internal invariants guarantee it to be safe (with a comment providing justification). We assume that everything external to the program (e.g., `docker` invocations) can and will fail for whatever reason (e.g., maybe the Docker daemon isn't running).

It's a little trickier to write the code this way, but it's worth it.

stepchowfun commented on src/run.rs:
```diff
@@ -95,12 +111,40 @@ pub fn image_ids() -> io::Result<HashSet<String>> {
+                    if line.is_empty() {
+                        let tab_index = line.find('\t').unwrap();
+                        let (image_id, date_str) = line.split_at(tab_index);
+                        Some((image_id.to_owned(), parse_docker_date(date_str.trim())))
```
Can we `trim` the `image_id` as well?

stepchowfun replied:
I understand the other `trim` is necessary because of the tab character, whereas the `image_id` `trim` is probably not needed. But it would help me sleep at night if we trim everything that comes from the output of an external program.

stepchowfun commented on src/run.rs:
```diff
@@ -95,12 +111,40 @@ pub fn image_ids() -> io::Result<HashSet<String>> {
+fn parse_docker_date(date_str: &str) -> Duration {
+    let date_fmt = "%Y-%m-%d %H:%M:%S %z";
+    // Chrono can't read the "EST", so remove it before parsing
+    let timezone_index = 25;
```
This hardcoded index seems brittle. Could we instead use something like `rsplitn` to remove it?

stepchowfun commented on src/run.rs:
```diff
@@ -95,12 +111,40 @@ pub fn image_ids() -> io::Result<HashSet<String>> {
+        warn!(
+            "Unable to parse CreatedAt: {}. Defaulting to current time",
+            date_str
+        );
```
Instead of logging a warning and defaulting to the current time, what do you think about just failing hard (e.g., by returning an `io::Error`)? Such a failure would then become an error for the `image_ids` function above.

I'm on the fence, but I'm thinking it might be better to fail loudly if something here goes wrong (so that it will be brought to my attention such that I can fix it)?

stepchowfun commented on src/run.rs:
```diff
@@ -95,12 +111,40 @@ pub fn image_ids() -> io::Result<HashSet<String>> {
+            "Unable to parse CreatedAt: {}. Defaulting to current time",
+            date_str
+        );
+        Local::now().timestamp()
```
I think this is correct, but `Utc::now().timestamp()` might be preferable because it avoids an unnecessary timezone calculation.

stepchowfun commented on src/run.rs:
```diff
@@ -95,12 +111,40 @@ pub fn image_ids() -> io::Result<HashSet<String>> {
+        );
+        Local::now().timestamp()
+    };
+    Duration::from_secs(timestamp.try_into().unwrap())
```
I have a slight preference for not converting the timestamp into an integer and then creating a duration out of that. The reason is that we lose some type safety: we have to rely on the fact that the integer represents a number of seconds since the UNIX epoch, which is in some sense a lower-level implementation detail of how timestamps are represented.

I think it could be better to be explicit about this calculation like we do here:

https://github.com/stepchowfun/docuum/blob/b246d638c0d99d0f40b06822685cc44dd307e055/src/run.rs#L233

(But instead of `SystemTime::now()`, maybe `Utc::now()` since we are in `chrono`-land here?)

Unfortunately it would be more verbose, though.

More importantly though: I prefer to avoid the `unwrap` if possible (and if not possible, justify it with a comment).

mac-chaffee replied:
Is it possible to convert from a `chrono::DateTime` to a `std::time::Duration` without an intermediate step as `i64`? Looks like the direct conversion from SystemTime -> Duration is unique to SystemTime.

What's the risk if we do that conversion to i64? Wouldn't the risk just be deleting images slightly less optimally if the system time doesn't increase monotonically? It wouldn't crash, would it?

mac-chaffee replied:
For context, this is what I have so far to remove that `unwrap()`, but it still has the type conversion:

```rust
    let dt = match DateTime::parse_from_str(&clean_date_str, "%Y-%m-%d %H:%M:%S %z") {
        Ok(dt) => dt,
        Err(e) => return Err(io::Error::new(io::ErrorKind::Other, e)),
    };
    let num_seconds = match u64::try_from(dt.timestamp()) {
        Ok(num_seconds) => num_seconds,
        Err(e) => return Err(io::Error::new(io::ErrorKind::Other, e)),
    };
    Duration::from_secs(num_seconds))
```

stepchowfun replied:
> Is it possible to convert from a `chrono::DateTime` to a `std::time::Duration` without an intermediate step as `i64`? Looks like the direct conversion from SystemTime -> Duration is unique to SystemTime.

I think `chrono::DateTime` has [`signed_duration_since`](https://docs.rs/chrono/0.4.13/chrono/struct.DateTime.html#method.signed_duration_since). Maybe we can use that? I'm a little bit confused though by the fact that it returns an `OldDuration`, and I'm not sure what that is. If my suggestion turns out to be too awkward to implement, we can go with your implementation (from your comment in this thread).

> What's the risk if we do that conversion to i64? Wouldn't the risk just be deleting images slightly less optimally if the system time doesn't increase monotonically? It wouldn't crash, would it?

It's not a correctness issue (I think the non-monotonicity concern applies equally to both cases?), it's just a (very) minor style issue. Going from a `DateTime` directly to a `Duration` seems (to me, at least) slightly better than doing an intermediate conversion to a plain integer, because with an integer we're no longer encoding our intent in the type system. It's a lower-level representation that offers no guarantees about how it's used.

In my mind, it's kind of like creating [newtypes](https://doc.rust-lang.org/stable/rust-by-example/generics/new_types.html) to represent different units like `Meters`, `Miles`, etc. They're all integers under the hood, but using a more specific type helps the type system catch mistake.

Let me know if that makes sense.

stepchowfun replied:
Just to clarify: I think for this particular code it doesn't matter that much, because the logic is simple enough that it's unlikely there will be a hidden mistake in it. But for more complicated logic it really does help to leverage the type system like this, so I prefer to just apply this kind of rigor to everything, no matter how trivial. But I understand this makes me an annoying code reviewer sometimes!

mdonoughe replied:
`OldDuration` is an alias for `chrono::Duration`. See https://docs.rs/chrono/0.4.13/chrono/index.html#duration

mac-chaffee replied:
Ahh yeah that works. I remember seeing that `signed_duration_since` function, but skipped it since I couldn't think of a nice way to pass `UNIX_EPOCH` into that function (until I discovered `DateTime::from(UNIX_EPOCH)` today)

stepchowfun commented on toast.yml:
```diff
@@ -94,7 +94,7 @@ tasks:
       cargo clippy --all-targets --all-features -- \
         --deny warnings --deny clippy::all --deny clippy::pedantic
-      cargo fmt --all -- --check
+      cargo fmt --all -- --check || echo "ERROR: Please correct the formatting errors above"
```
Is this because it was difficult to understand the output? I wonder if we should do this for all the lint checks (or none of them).

Either way, a few minor suggestions:
- `echo` to standard error rather than standard out (e.g., `echo '...' 1>&2`)
- Punctuate the sentence.
- Use single quotes unless we need the interpolation capability of double quotes. I know this is nitpicky, but I find that this practice avoids mistakes in which Bash interprets some part of the string that you didn't intend it to (though here I think it's inconsequential).

mac-chaffee replied:
Yes, when `cargo fmt` fails, it just prints a`diff` and I didn't know why or even what command was triggering that.
Whereas `clippy` has some explanation like `error: unneeded return statement`, so I knew how to address that.

stepchowfun reviewed (commented):
This is looking really good. I left some minor comments. The most important feedback is to avoid `unwrap` as much as possible, unless we can be absolutely sure that it's safe (with justification provided in comments). If there is even the remotest possibility that an `unwrap` will `panic`, we should return an error instead so the caller can decide what to do.

mac-chaffee commented:
This is probably the highest quality code review I've ever received, thanks @stepchowfun! I'll get started on addressing these comments

mac-chaffee commented:
Still trying to find time to pick another fight with the Rust compiler to address these changes. If it turns out #75 is done and ready to be merged before this PR, just letting you know I'd be okay with fixing the resulting merge conflicts if need-be 👍

mac-chaffee committed:
Send lint error output to stderr

mac-chaffee committed:
Improve error handling for parse_docker_date

mac-chaffee committed:
Move all date-parsing code into parse_docker_date

mac-chaffee commented:
I resolved the comments I think I've addressed, feel free to re-comment since everything's shifted around now. Mainly did the following:

* Replaced the `unwrap`s with either `ok_or_else` or `match`. Handling errors seems to be a common noob trap, so more guidance is always appreciated
* Replaced the `map` chain with a for-loop. It hurt me to abandon the clever iterator tricks, but I figured a for loop would help readability, especially since things seem to get messy when the `map`'d function is fallible. Open to other advice here too.

stepchowfun commented on src/run.rs:
```diff
@@ -88,17 +97,47 @@ pub fn image_ids() -> io::Result<HashSet<String>> {
+                match parse_docker_date(&date_str) {
+                    Ok(duration) => images.insert(image_id.trim().to_owned(), duration),
+                    Err(e) => return Err(e),
+                };
```
Instead of doing a pattern match here, would the `?` shorthand syntax work? Something like:

```rust
images.insert(image_id.trim().to_owned(), parse_docker_date(&date_str)?);
```

stepchowfun reviewed (commented):
This is looking good to me. Would you mind rebasing now that the other PR has been merged? (Thanks for offering to do that.)

stepchowfun commented on src/run.rs:
```diff
@@ -4,13 +4,15 @@ use crate::{
 use scopeguard::guard;
 use serde::{Deserialize, Serialize};
 use std::{
+    collections::HashMap,
```
I think we can DRY this up a tiny bit: `collections::{HashMap, HashSet}`.

stepchowfun commented on src/run.rs:
```diff
@@ -88,17 +97,47 @@ pub fn image_ids() -> io::Result<HashSet<String>> {
+                    continue;
+                }
+                let tab_index = line.find('\t').ok_or_else(|| {
+                    io::Error::new(io::ErrorKind::Other, "Failed to split image ID and date")
```
For consistency with the existing error messages, can we add a period at the end of the string?

stepchowfun commented on src/run.rs:
```diff
@@ -88,17 +97,47 @@ pub fn image_ids() -> io::Result<HashSet<String>> {
+    // Chrono can't read the "EST", so remove it before parsing
+    let clean_date_str =
+        date_str.trim().rsplitn(2, ' ').last().ok_or_else(|| {
+            io::Error::new(io::ErrorKind::Other, "Failed to remove timezone string")
```
Same comment as above regarding punctuating error messages.

mac-chaffee committed:
Add dryness and punctuation

mac-chaffee commented:
I figured I'd have to throw away a lot of code when #75 was merged, but I learned a lot anyway!

Also for some reason I am now unable to run docuum locally after compiling. Filed a bug: #78

mac-chaffee commented:
So @stepchowfun do you have a preference on what to do with this PR now that #75 was reverted?

If there's a quick path to getting Bollard back in, we could just leave this PR until that point.
Otherwise, I could try to recover those commits I squashed and go back to parsing CreatedAt from the CLI

stepchowfun commented on src/run.rs:
```diff
@@ -36,7 +36,7 @@ pub async fn image_id(docker: &Docker, image: &str) -> io::Result<String> {
 
 // Ask Docker for the IDs of all the images.
-pub async fn image_ids(docker: &Docker) -> io::Result<HashSet<String>> {
+pub async fn image_ids(docker: &Docker) -> io::Result<HashMap<String, Duration>> {
```
What do you think about renaming this function to `image_ids_and_creation_timestamps`? It's kind of verbose, but the current name feels incomplete now.

stepchowfun replied:
(If this suggestion resonates with you, we can also rename the `image_ids` variable defined on line 169 accordingly.)

stepchowfun commented on src/run.rs:
```diff
@@ -50,8 +50,15 @@ pub async fn image_ids(docker: &Docker) -> io::Result<HashSet<String>> {
-    Ok(output.into_iter().map(|image| image.id).collect())
+    for image in output {
+        match image.created.try_into() {
+            Ok(unsigned) => images.insert(image.id, Duration::from_secs(unsigned)),
```
This is totally subjective, but I think the `unsigned` variable could have a better name. Maybe `created` to match the name of where it comes from? (I'm not super fond of `created`, but consistency is more important than my personal preferences.)

stepchowfun commented on src/run.rs:
```diff
@@ -50,8 +50,15 @@ pub async fn image_ids(docker: &Docker) -> io::Result<HashSet<String>> {
+    for image in output {
+        match image.created.try_into() {
+            Ok(unsigned) => images.insert(image.id, Duration::from_secs(unsigned)),
+            Err(e) => return Err(io::Error::new(io::ErrorKind::Other, e))
```
Elsewhere in this file we use the name `error` rather than `e` when pattern matching on an `Err`. Can we do that here too? (Alternatively we could change all the other ones. I don't feel strongly about it, as long as we're consistent.)

stepchowfun reviewed (commented):
@mac-chaffee If I understand the situation correctly, I don't think we'll be able to bring back Bollard any time soon. If it's a bug in Docker, I guess we just need to wait for all the affected Docker versions (which includes the latest current version) to fall out of use?

I left some comments on the current version of the PR, but unfortunately I think we'll have to go back to the old way (parsing the output of the Docker CLI) so my comments might be pointless. I'm sorry you had to rebase your PR only to find out that we have to revert back to the old strategy after all. That's a big bummer!

Let me know if you are able to recover your old commits to restore the PR to how it was before the rebasing. I think it was essentially ready to ship, or at least close to it.

mdonoughe commented:
We shouldn't need to wait that long. The docker team changed the documentation to match the current behavior so we just need Bollard to pick that up.

stepchowfun commented:
@mdonoughe Oh so there is no behavior change on Docker's side? That's kind of unfortunate, because the behavior is weirdly inconsistent (integer vs. timestamp string) depending on whether BuildKit is enabled, right? Should Bollard have to deal with that?

mdonoughe commented:
When you're using BuildKit there's an extra section specifically about BuildKit cache data and that's where the problem is.

stepchowfun commented:
Ah, that makes sense. Thank you for clarifying! In that case, I think this can go either way: wait for the Bollard upstream fix (do they know about the issue?), bring back @mdonoughe's change, and then land this PR, or alternatively restore this PR to how it was before Bollard, land it, and then land the Bollard change separately when the time is right.

Either way is fine with me, but I imagine it could be faster to update this PR and land it rather than waiting on Bollard.

Fortunately, neither of these changes has any timeline requirement on my end, so it's up to y'all to decide how you'd like to proceed.

mdonoughe commented:
I created fussybeaver/bollard#89 earlier today.

mac-chaffee committed:
Rename image_ids -> image_ids_and_creation_timestamps

mac-chaffee commented:
Did some (surprisingly easy) `git reflog` surgery and got the code back! The old code is at https://github.com/mac-chaffee/docuum/tree/BACKUP2 but the conversion of this change back into using bollard should be straight-forward.

mac-chaffee committed:
Merge branch 'master' into master

stepchowfun commented:
Nice! Looks good.
